openvdb/openvdb/tree/ValueAccessor.h

```diff
@@ -132,3134 +179,767 @@ class ValueAccessorBase
+                    assert(mBuffer);
+                    return &(mBuffer[LeafNodeT::coordToOffset(xyz)]);
+                }
+                else {
+                    auto node = mNodes.template get<Idx>();
+                    assert(node);
+                    return &(node->getValueAndCache(xyz, *this));
+                }
+            });
+    }
+
+    /// @brief Return the active state of the voxel at the given coordinates.
+    /// @param xyz  The index space coordinate to query
+    bool isValueOn(const Coord& xyz) const
+    {
+        return this->evalFirstCached(xyz, [&](const auto node) -> bool {
+                assert(node);
+                return node->isValueOnAndCache(xyz, *this);
+            });
     }
 
-    /// Return the active state of the voxel at the given coordinates.
-    bool isValueOn(const Coord& xyz) const { LockT lock(mMutex); return mCache.isValueOn(xyz); }
-
-    /// Return the active state of the voxel as well as its value
+    /// @brief Return the active state of the value at a given coordinate as
+    ///   well as its value.
+    /// @param xyz    The index space coordinate to query
+    /// @param value  The value to get
     bool probeValue(const Coord& xyz, ValueType& value) const
     {
-        LockT lock(mMutex);
-        return mCache.probeValue(xyz,value);
-    }
-
-    /// Return the tree depth (0 = root) at which the value of voxel (x, y, z) resides,
-    /// or -1 if (x, y, z) isn't explicitly represented in the tree (i.e., if it is
-    /// implicitly a background voxel).
+        return this->evalFirstCached(xyz, [&](const auto node) -> bool
+            {
+                using NodeType = std::remove_pointer_t<decltype(node)>;
+                assert(node);
+
+                if constexpr(IsLeafAndBypassLeafAPI<NodeType>) {
+                    assert(mBuffer);
+                    const auto offset = LeafNodeT::coordToOffset(xyz);
+                    value = mBuffer[offset];
+                    return node->isValueOn(offset);
+                }
+                else {
+                    return node->probeValueAndCache(xyz, value, *this);
+                }
+            });
+    }
+
+    /// @brief Return the tree depth (0 = root) at which the value of voxel
+    ///   (x, y, z) resides, or -1 if (x, y, z) isn't explicitly represented in
+    ///   the tree (i.e., if it is implicitly a background voxel).
+    /// @note  This is the inverse of the node LEVEL (where the RootNode level
+    ///   is the highest in the tree).
+    /// @param xyz  The index space coordinate to query
     int getValueDepth(const Coord& xyz) const
     {
-        LockT lock(mMutex);
-        return mCache.getValueDepth(xyz);
+        return this->evalFirstCached(xyz, [&](const auto node) -> int
+            {
+                using NodeType = std::remove_pointer_t<decltype(node)>;
+                assert(node);
+
+                if constexpr(std::is_same<RootNodeT, NodeType>::value) {
```

Comment:
There's a similar pattern to this used in the `isCached(const Coord& xyz)` method which fails on GCC9. Do you have any insight as to why this one succeeds and the other fails? Is it worth perhaps always pre-evaluating the conditional logic of these constexpr expressions given that there may be other compilers we're not testing against that fail in a similar way?

Reply:
I've had a look at this and don't think I need to change it. The stuff that is crashing takes the form:

```
                using NodeType = typename NodeLevelList::template Get<Idx>;
                // @warning  Putting this exp in the if statement crashes GCC9
                constexpr bool IsRoot = std::is_same<RootNodeT, NodeType>::value;
```

Basically a `TypeList` access followed by an `is_same`.

The stuff remaining takes the form:

```
                using NodeType = std::remove_pointer_t<decltype(node)>;
                assert(node);
                constexpr bool IsRoot = std::is_same<RootNodeT, NodeType>::value;
```

Where `node` is the lambda runtime argument. These are two very different instantiations so it makes sense why we'd see a crash in one and not the other. I will leave this as is for now.